### PR TITLE
Use "json-seq" for long-lived connections without need for polling

### PR DESCRIPTION
**Pull request by olivercalder**
By using `json-seq`, long-lived connections can be opened to the snapd requests API so that prompt clients will be automatically notified when a new request comes in.  This can be done by passing the `follow=true` option in the query arguments.  In this case, it is no longer necessary for snapd to send information to a client socket in order to tell the client to query the snapd API for request information.

Thus, registering a client socket with snapd is also no longer required. Instead, snapd must be able to identify whether a client should be able to connect to the snapd request endpoint, and only allow appropriate clients to be able to connect.

If a client, such as a settings application which is designed to modify stored decisions, wants to monitor changes to stored decisions without repeatedly querying the snapd decision API, it can use the same `follow=true` option in the query to open a long-lived connection in the same way.

### CODE REVIEW

**Review by robert-ancell — APPROVED** [review]
Looks good to me - @mvo5 does this match the pattern you expect in snapd?

**Comment by tim-hm** [comment]
Event-based is a good idea imo and there is a lot of prior art for how these systems work. A few characteristics to discuss:

1. Does json-seq support bidirectional communications like web sockets? Ie would the response use the same channel as the stream?

    a. One interesting way we could implement a bidirectional channel is ensuring exactly 1 prompting client can register

2. What delivery methodology would we use? For example, at-most-once, at-least-once, exactly-once, best-effort.

 - This is related to  1a. If we allowed multiple prompting agents to connect, would we let them all consume the message? If we allow at most 1 to connect, what do we do with failed deliveries?

3. For the channel(s) do we use topic based (ie one for prompts, one for changes, etc) or a single channel for everything?

Finally, I think it would be useful to understand before pursing this:

- @mvo5 @olivercalder How easily could an endpoint like this make its way into snapd? 
- @robert-ancell Are there implementation challenges for say a web socket client in c?

For example, perhaps we need to crack on with a simpler restful approach to get an preview build in place before we take on too much complexity.

**Comment by robert-ancell** [comment]
>     1. Does json-seq support bidirectional communications like web sockets? Ie would the response use the same channel as the stream?
>        a. One interesting way we could implement a bidirectional channel is ensuring exactly 1 prompting client can register

I don't know if it matters if more than one prompting client exists, only that the clients are allowed to do prompting. For example, you could imagine a case where a prompt would show in a pop-up window and also on your phone. The user would then choose to respond to either prompt and as long as both clients knew the prompt had been responded they could reflect that in their UIs.
 
>     2. What delivery methodology would we use? For example, at-most-once, at-least-once, exactly-once, best-effort.

exactly-once I would think.

>     * This is related to  1a. If we allowed multiple prompting agents to connect, would we let them all consume the message? If we allow at most 1 to connect, what do we do with failed deliveries?

Yes, they would get the same messages, if two clients tried to respond to the prompt at the same time the second one would receive an error. But it would also receive the result of the prompt response from the first client.

>     3. For the channel(s) do we use topic based (ie one for prompts, one for changes, etc) or a single channel for everything?

I think @mvo5 needs to weigh in on this but it seems the convention is topic based, as there exists /v2/logs?follow=true, and a proposal for /v2/changes?follow=true.

> Finally, I think it would be useful to understand before pursing this:
> 
>     * @mvo5 @olivercalder How easily could an endpoint like this make its way into snapd?
> 
>     * @robert-ancell Are there implementation challenges for say a web socket client in c?

There was issues using websockets previously with libsoup2, as it didn't support Unix sockets and we had to do a bit of manual HTTP implementation to make it work. This is fixed in libsoup3 (I made the upstream change), so I think websockets would work, though I've never used them so I can't say for sure. If we are required to support LTS releases that are still using libsoup2 then we will have issues.

I haven't seen the need to go full websockets as the above solution seems sufficient and simpler.

**Comment by olivercalder** [comment]
> I don't know if it matters if more than one prompting client exists, only that the clients are allowed to do prompting. For example, you could imagine a case where a prompt would show in a pop-up window and also on your phone. The user would then choose to respond to either prompt and as long as both clients knew the prompt had been responded they could reflect that in their UIs.

> Yes, they would get the same messages, if two clients tried to respond to the prompt at the same time the second one would receive an error. But it would also receive the result of the prompt response from the first client.

I really like this approach @robert-ancell .  My only concern is there could be race conditions if both clients respond before snapd has a chance to handle the first response and know to reply to the second with an error and the previous decision, but if it handles requests sequentially this should be fine.


>>    3. For the channel(s) do we use topic based (ie one for prompts, one for changes, etc) or a single channel for everything?

> I think @mvo5 needs to weigh in on this but it seems the convention is topic based, as there exists /v2/logs?follow=true, and a proposal for /v2/changes?follow=true.

@tim-hm I think having different endpoints for requests and previously-stored decisions (including listening for changes) is helpful, where prompt clients would usually connect on the requests channel and GNOME-control-center type applications would connect on the decisions channel. Please correct me though if I'm misunderstanding.

> @mvo5 @olivercalder How easily could an endpoint like this make its way into snapd?

From a technical perspective this is not hard at all, we can simply add new endpoints alongside existing ones as defined in the [snapd REST API documentation](https://snapcraft.io/docs/snapd-rest-api). Changes would definitely need review/approval by Samuele. However, I just met with Michael and he suggested I start opening small PRs to snapd with the API endpoints, so I may get this process rolling.

**Comment by olivercalder** [comment]
In my view, the remaining question before merging this PR is: Can multiple clients (either for different users or for the same user) be connected to the same snapd API endpoint at the same time (via "json-seq"), and can snapd differentiate between client UIDs and only send requests to the correct client?

The alternative (inspired by @tim-hm in https://github.com/olivercalder/snapd-prompt-api-brainstorming/pull/3) would be to have something like `/v2/prompting/{UID}/**`, so everything would be done per-user.

I think @mvo5 might know best about the former in particular. I'm open to either approach.

**Comment by tim-hm** [comment]
1. Do we need to answer the multi-client question at this stage ... ie can we get to a working alpha and then reassess?

2. In terms of ws vs json-seq ... I think library support/tooling should win out there and I'm not a go developer so just throwing approaches out that I've used elsewhere where the support was mature for ws. Again lots of prior art with approaches like CQRS which can be helpful for inspiration -- or more likely that your all over this already and don't need it!

3. Wrt `/v2/promptin/{UID}/**`, my worry with that approach is how are we preventing `bob` from approving prompts for `sally`? For example, if the socket is at `/run/user/bob/socket` then its easier to restrict access to UID bob.

> I just met with Michael and he suggested I start opening small PRs to snapd with the API endpoints, so I may get this process rolling.

4. My feel is that the minimal we need is the "prompt stream" and the response endpoint but useful for @robert-ancell to weigh in here as well.

**Comment by olivercalder** [comment]
>     1. Do we need to answer the multi-client question at this stage ... ie can we get to a working alpha and then reassess?

Yes, that's totally reasonable. I think it's worth figuring this out before it lands in snapd:master, but for now this is definitely sufficient.

>     2. In terms of ws vs json-seq ... I think library support/tooling should win out there and I'm not a go developer so just throwing approaches out that I've used elsewhere where the support was mature for ws. Again lots of prior art with approaches like CQRS which can be helpful for inspiration -- or more likely that your all over this already and don't need it!

I don't think I've heard of "ws" before... what is it?

Edit: WebSocket, (ws:// instead of http://) which uses port 80 and 443, but allows full-duplex connections and streaming. This sounds great, and seems like it's exactly what we're looking for (and sounds just like json-seq). It sounds like the benefit of WebSockets would be allowing responses back to snapd over the same connection opened by the client -- is that something json-seq also supports?

>     3. Wrt `/v2/promptin/{UID}/**`, my worry with that approach is how are we preventing `bob` from approving prompts for `sally`? For example, if the socket is at `/run/user/bob/socket` then its easier to restrict access to UID bob.

My understanding is having separate endpoints for each UID solves this problem (if it is a problem). That is (AFAIK), snapd can restrict who can connect to sockets, so if we only allow client(s) with the correct UID to connect to the corresponding UID-specific snapd socket, then snapd can correctly send prompt requests for that UID to that socket, and the correct client(s) can then handle the request.

As opposed to a single socket for all clients, where then snapd needs to have some way of only sending requests to particular connected clients on a shared socket, which may not be possible. Again, this is very much "AFAIK", so anyone please correct me if I'm wrong.

>     4. My feel is that the minimal we need is the "prompt stream" and the response endpoint but useful for @robert-ancell to weigh in here as well.

This sounds good to me. By "prompt stream", do you mean the `/v2/prompting/requests**` endpoint, and by "response endpoint", does `POST`ing to `/v2/prompting/requests/{id}` qualify?

**Comment by olivercalder** [comment]
I'm going to merge this PR (and https://github.com/olivercalder/snapd-prompt-api-brainstorming/pull/3) and get started on a new one which assumes both.

The open questions:

- json-seq vs ws
- shared or UID-specific endpoints

can be resolved with future PRs in the future, if necessary. Thanks again for all the feedback @robert-ancell and @tim-hm !

**Pull request merged**
